sources/ghostdog/src/main.rs

```diff
@@ -319,6 +368,294 @@ lazy_static! {
+                continue;
+            }
+
+            let parts: Vec<&str> = line.split(":").collect();
+            if parts.len() != 2 {
+                continue;
+            }
+            let key = parts[0].trim();
+            let value = parts[1].trim();
+
+            // Match the key to the left and set the right
+            match key {
+                "CA type" => ca.ca_type = value.to_string(),
+                "Number of ports" => ca.num_ports = value.parse()?,
+                "Firmware version" => ca.firmware_version = value.to_string(),
+                "Hardware version" => ca.hardware_version = value.parse()?,
+                "Node GUID" => ca.node_guid = value.to_string(),
+                "System image GUID" => ca.system_image_guid = value.to_string(),
+                // Port fields
+                "State" => {
+                    if let Some(port) = &mut current_port {
+                        port.state = value.to_string();
+                    }
+                }
+                "Physical state" => {
+                    if let Some(port) = &mut current_port {
+                        port.physical_state = value.to_string();
+                    }
+                }
+                "Rate" => {
+                    if let Some(port) = &mut current_port {
+                        port.rate = value.parse()?;
+                    }
+                }
+                "Base lid" => {
+                    if let Some(port) = &mut current_port {
+                        port.base_lid = value.parse()?;
+                    }
+                }
+                "LMC" => {
+                    if let Some(port) = &mut current_port {
+                        port.lmc = value.parse()?;
+                    }
+                }
+                "SM lid" => {
+                    if let Some(port) = &mut current_port {
+                        port.sm_lid = value.parse()?;
+                    }
+                }
+                "Capability mask" => {
+                    if let Some(port) = &mut current_port {
+                        port.capability_mask = value.to_string();
+                    }
+                }
+                "Port GUID" => {
+                    if let Some(port) = &mut current_port {
+                        port.port_guid = value.to_string();
+                    }
+                }
+                "Link layer" => {
+                    if let Some(port) = &mut current_port {
+                        port.link_layer = value.to_string();
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        // After all the line parsing, put the last port into the Vec.
+        if let Some(port) = current_port {
+            ca.ports.push(port);
+        }
+        Ok(ca)
+    }
+}
+
+/// Finds the device name using the nvme-cli
+fn get_ibstat_for_device(device: OsString) -> Result<InfinibandDevice> {
+    let device_name = device.into_string().unwrap();
```

Comment:
snafu really was fighting me here because the Result is actually of `Result<String, OsString>` which is complicated to work with.

Reply:
[`Command::args` accepts as `AsRef<OsStr>`](https://doc.rust-lang.org/std/process/struct.Command.html#method.args) (which `&OsString` implements), so I think you actually don't need to do a UTF-8 check here -- you can just pass the `device_name` directly to `args()`.

---

In case I'm wrong:

The reason this is like this is that `into_string()` takes ownership of `device`; however, it returns an `Err` if the string is not valid UTF-8.

If you want to subsequently print an error message that includes the non-UTF-8 string, you'll need access to it after the call to `into_string()`. If the function didn't give you the string back, you'd have to clone it.

I think what you'll basically want to do is something like:

```suggestion
    let device_name = device.into_string().map_err(|device_name| {
        // `display` for this mentions that the device name isn't valid UTF-8
        error::Error::DeviceNameEncoding {
            device_name
        }
    });
```

Reply:
I'm not sure I got myself into a place where I was doing the extra `Into` but without it works fine.